vendors/nordic/boards/nrf52840-dk/ports/ota_pal_for_aws/ota_pal.c

```diff
@@ -0,0 +1,734 @@
+        }
+    }
+
+    /* Read the old image */
+    /* Right now we always boot from the first bank */
+    /* TODO: Support boot from the second bank */
+    ImageDescriptor_t * old_descriptor = ( ImageDescriptor_t * ) ( otapalFIRST_BANK_START );
+    ImageDescriptor_t new_descriptor;
+
+    /* Check if the correct image is located at the beginning of the bank */
+    if( ( memcmp( old_descriptor->pMagick, pcOTA_PAL_Magick, otapalMAGICK_SIZE ) != 0 ) && ( eState == OtaImageStateAccepted ) )
+    {
+        xStatus = OtaPalCommitFailed;
+    }
+
+    if( xStatus == OtaPalSuccess )
+    {
+        memcpy( &new_descriptor, old_descriptor, sizeof( new_descriptor ) );
+
+        switch( eState )
+        {
+            case OtaImageStateAccepted:
+
+                if( old_descriptor->usImageFlags == otapalIMAGE_FLAG_COMMIT_PENDING )
+                {
+                    new_descriptor.usImageFlags = otapalIMAGE_FLAG_VALID;
+                }
+                else
+                {
+                    new_descriptor.usImageFlags = otapalIMAGE_FLAG_INVALID;
+                }
+
+                break;
+
+            case OtaImageStateRejected:
+            case OtaImageStateAborted:
+
+                if( old_descriptor->usImageFlags == otapalIMAGE_FLAG_COMMIT_PENDING )
+                {
+                    new_descriptor.usImageFlags = otapalIMAGE_FLAG_INVALID;
+                }
+
+                break;
+
+            case OtaImageStateTesting:
+                break;
+
+            default:
+                xStatus = OtaPalBadImageState;
+                break;
+        }
+    }
+
+    if( xStatus == OtaPalSuccess )
+    {
+        /* We don't wont to do anything with flash if it would leave it in the same state as it were */
+        if( ( new_descriptor.usImageFlags != old_descriptor->usImageFlags ) && ( eState != OtaImageStateTesting ) )
+        {
+            xStatus = prvErasePages( otapalFIRST_BANK_START, otapalFIRST_BANK_START + otapalDESCRIPTOR_SIZE );
+
+            if( xStatus == OtaPalSuccess )
+            {
+                xStatus = prvWriteFlash( otapalFIRST_BANK_START, sizeof( new_descriptor ), ( uint8_t * ) &new_descriptor );
+            }
+        }
+    }
+
+    return OTA_PAL_COMBINE_ERR( xStatus, 0 );
+}
+/*-----------------------------------------------------------*/
+
+OtaPalImageState_t otaPal_GetPlatformImageState( OtaFileContext_t * const pFileContext )
+{
+    OtaPalImageState_t xImageState = OtaPalImageStateInvalid;
+
+    ImageDescriptor_t * descriptor = ( ImageDescriptor_t * ) ( otapalFIRST_BANK_START );
+
+    switch( descriptor->usImageFlags )
+    {
+        case otapalIMAGE_FLAG_COMMIT_PENDING:
+            xImageState = OtaPalImageStatePendingCommit;
+            break;
+
+        case otapalIMAGE_FLAG_VALID:
+            xImageState = OtaPalImageStateValid;
+            break;
+
+        case otapalIMAGE_FLAG_INVALID:
+            xImageState = OtaPalImageStateInvalid;
+            break;
```

Comment:
Are these the only three possible values? Should we have a default case for error handling?